packages/tizen_rpc_port/lib/src/rpc_port_method_channel.dart

```diff
@@ -0,0 +1,200 @@
+import 'dart:typed_data';
+import 'package:flutter/services.dart';
+import 'package:tizen_log/tizen_log.dart';
+
+import 'port.dart';
+import 'proxy_base.dart';
+
+const String _logTag = 'RpcPortMethodChannel';
+
+/// An implementation of [RpcPortPlatform] that uses method channels.
+class MethodChannelRpcPort {
+  static final MethodChannelRpcPort instance = MethodChannelRpcPort();
+
+  /// The method channel used to interact with the native platform.
+  final MethodChannel _channel = const MethodChannel('tizen/rpc_port');
+
+  final Map<String, Stream<dynamic>> _streams = <String, Stream<dynamic>>{};
+
+  String _createKey(String appid, String portName) => '$appid/$portName';
+
+  Future<void> stubCreate(String portName) async {
+    final Map<String, String> args = <String, String>{'portName': portName};
+    return await _channel.invokeMethod('stubCreate', args);
+  }
+
+  Future<void> stubDestroy(String portName) async {
+    final Map<String, String> args = <String, String>{'portName': portName};
+    return await _channel.invokeMethod('stubDestroy', args);
+  }
+
+  Future<Stream<dynamic>> proxyConnect(ProxyBase proxy) async {
+    final String key = _createKey(proxy.appid, proxy.portName);
+    if (_streams.containsKey(key)) {
+      return _streams[key]!;
+    }
+
+    final EventChannel eventChannel = EventChannel('tizen/rpc_port_proxy/$key');
+    _streams[key] = eventChannel.receiveBroadcastStream();
```

Comment:
I don't think having event channel for each (appid, portName) is such a good idea. You can create a single event channel for proxies and provide each stream that filters based on appid and portName. For example, see the [camera_plugin](https://github.com/flutter/plugins/blob/2cbb314d99d800cb4e1229ce6b304755303bbd64/packages/camera/camera_platform_interface/lib/src/method_channel/method_channel_camera.dart#L59-L61).

By doing this you can remove a lot of event channel creation and maintenance logic both in Dart and C++.

For example:

In `rpc_port_method_channel.dart`:

```dart
enum RpcEventType {
  connected,
  disconnected,
  rejected,
  dataReceived,
}

/// Proxy event from platform.
class ProxyEvent {
  /// Creates an instance of [ProxyEvent].
  ProxyEvent({
    required this.event,
    required this.appid,
    required this.portName,
    this.parcel,
  });

  /// The type of event occurred.
  final RpcEventType event;

  /// The appid of the proxy where event occurred.
  final String appid;

  /// The port name of the proxy where event occurred.
  final String portName;

  /// The received parcel data on [RpcEventType.dataReceived] event.
  ///
  /// Is `null` for other type of events.
  final Parcel? parcel;
}

class MethodChannelRpcPort {
  static final MethodChannelRpcPort instance = MethodChannelRpcPort();

  final EventChannel _proxyEventChannel =
      const EventChannel('tizen/rpc_port/proxy');

  Stream<ProxyEvent>? _proxyEvents;

  Stream<ProxyEvent> get _proxyEventsInternal {
    _proxyEvents ??=
        _proxyEventChannel.receiveBroadcastStream().map((dynamic event) {
      final Map<String, Object> map = (event as Map).cast<String, Object>();
      final String eventTypeString = map['event'] as String;
      RpcEventType? eventType;
      for (final RpcEventType value in RpcEventType.values) {
        if (value.name == eventTypeString) {
          eventType = value;
          break;
        }
      }
      if (eventType == null) {
        throw PlatformException(code: 'Undefined event type.');
      }
      final String appid = map['receiver'] as String;
      final String portName = map['portName'] as String;
      Parcel? parcel;
      if (eventType == RpcEventType.dataReceived) {
        final Uint8List rawData = map['rawData'] as Uint8List;
        parcel = Parcel.fromRaw(rawData);
      }
      return ProxyEvent(
        event: eventType,
        appid: appid,
        portName: portName,
        parcel: parcel,
      );
    });
    return _proxyEvents!;
  }

  Stream<ProxyEvent> proxyEventsFor(ProxyBase proxy) {
    return _proxyEventsInternal.where((ProxyEvent event) =>
        event.appid == proxy.appid && event.portName == proxy.portName);
  }
}
```

And in `proxy_base.dart`:

```dart
  ...

  /// Connects to the stub.
  Future<void> connect() async {
    if (_connected) {
      throw Exception('Proxy $_appid/$_portName already requested to stub');
    }

    await MethodChannelRpcPort.instance.proxyConnect(this);
    _listenForEvents();
  }

  /// Connects to the stub synchronously.
  Future<void> connectSync() async {
    if (_connected) {
      throw Exception('Proxy $_appid/$_portName already requested to stub');
    }

    await MethodChannelRpcPort.instance.proxyConnectSync(this);
    _listenForEvents();
  }

  void _listenForEvents() {
    _streamSubscription = MethodChannelRpcPort.instance
        .proxyEventsFor(this)
        .listen((ProxyEvent event) async {
      if (event.event == RpcEventType.connected) {
        await onConnectedEvent(appid, portName);
        _connected = true;
      } else if (event.event == RpcEventType.disconnected) {
        await onRejectedEvent(appid, portName);
        _streamSubscription?.cancel();
        _streamSubscription = null;
        _connected = false;
      } else if (event.event == RpcEventType.rejected) {
        await onRejectedEvent(appid, portName);
        _streamSubscription?.cancel();
        _streamSubscription = null;
        _connected = false;
      } else {
        await onReceivedEvent(appid, portName, event.parcel!);
      }
    });
  }

  ...
```

This can be done similarly for stub events as well.

P.S. I haven't built it so there might be some runtime error during casting.

Reply:
Thanks for this comment first.

The onConnectedEvent, onDisconnectedEvent, onRejectedEvent and onReceivedEvent method should be called for each request. When connected to two or more stubs, the proxy should receive each event method only for the target appid.
This implementation is similar to [https://github.com/flutter-tizen/plugins/blob/master/packages/messageport/lib/src/messageport_manager.dart](url).


[Question] Is there any problem if we create multiple event channels?
If it's, we will change the implementation as you mentioned.

Reply:
I think caching `Stream` objects in `_streams` is prone to bugs, but if you want to leave the current implementation please remove appropriate stream from `_streams` on `proxyDestroy`.

Reply:
To prevent bugs, I think your suggestion is better than current implementation.
@upple, How do you think? If you agree with me, please change the implementation to the suggested implementation.

Reply:
I added stream remove logic.

Reply:
The newly added `removeStream` exposes internal stream caching of `MethodChannelRpcPort`, this is not desired. I was talking about `_streams.remove(key)` in `proxyDestroy`, which you've already added.

Reply:
If what you removing it is a better choice, I'll erase it
Will this part be okay if I do that?

Reply:
Yes, please remove `removeStream` method from `MethodChannelRpcPort`.

Reply:
I removed it.
Thanks :)